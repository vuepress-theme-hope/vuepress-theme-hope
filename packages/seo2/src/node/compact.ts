import { createConverter } from "vuepress-shared/node";

import type { SeoOptions } from "./options.js";

/** @deprecated */
export const convertOptions = (
  options: SeoOptions & Record<string, unknown>,
): void => {
  const { droppedLogger } = createConverter("seo");

  droppedLogger({
    options,
    old: "seo",
    new: "ogp",
  });
  droppedLogger({
    options,
    old: "customMeta",
    new: "customHead",
  });
};
