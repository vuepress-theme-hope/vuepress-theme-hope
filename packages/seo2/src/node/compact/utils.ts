import { logger } from "../utils.js";

export const droppedLogger = (
  options: Record<string, unknown>,
  droppedOption: string,
  hint = "",
  newOption = ""
): void => {
  if (droppedOption in options) {
    logger.error(
      `"${droppedOption}" is removed${
        newOption
          ? `, please use ${newOption} instead.`
          : " and no longer supported"
      }${hint ? `\n${hint}` : ""}`
    );

    if (!newOption) delete options[droppedOption];
  }
};
