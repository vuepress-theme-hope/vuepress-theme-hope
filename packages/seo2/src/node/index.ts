import { seoPlugin } from "./plugin";

export * from "./plugin";
export * from "../shared";

export default seoPlugin;
