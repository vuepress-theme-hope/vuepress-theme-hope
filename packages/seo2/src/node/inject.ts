import type { HeadConfig } from "@vuepress/core";
import type { ArticleSeoContent, SeoContent, SeoOptions } from "../shared";

interface MetaOptions {
  name: string;
  content: string;
  attribute?: string;
}

const addMeta = (
  meta: HeadConfig[],
  {
    name,
    content,
    attribute = ["article:", "og:"].some((type) => name.startsWith(type))
      ? "property"
      : "name",
  }: MetaOptions
): void => {
  if (content) meta.push(["meta", { [attribute]: name, content }]);
};

export const appendSEO = (
  head: HeadConfig[],
  content: SeoContent,
  options: SeoOptions
): void => {
  for (const property in content)
    switch (property) {
      case "article:tag":
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        (content as ArticleSeoContent)["article:tag"]!.forEach((tag: string) =>
          addMeta(head, { name: "article:tag", content: tag })
        );
        break;
      case "og:locale:alternate":
        content["og:locale:alternate"].forEach((locale: string) => {
          if (locale !== content["og:locale"])
            addMeta(head, { name: "og:locale:alternate", content: locale });
        });
        break;
      default:
        if (content[property as keyof SeoContent] as string)
          addMeta(head, {
            name: property,
            content: content[property as keyof SeoContent] as string,
          });
    }

  if (options.restrictions)
    addMeta(head, {
      name: "og:restrictions:age",
      content: options.restrictions,
    });

  if (options.twitterID)
    addMeta(head, { name: "twitter:creator", content: options.twitterID });
};
