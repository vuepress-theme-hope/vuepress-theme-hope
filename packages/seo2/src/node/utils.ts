import matter from "gray-matter";
import {
  isLinkHttp,
  removeEndingSlash,
  removeLeadingSlash,
} from "@vuepress/shared";
import { Logger, isAbsoluteUrl, isUrl } from "vuepress-shared";

import type { App, SiteLocaleConfig } from "@vuepress/core";
import type { ExtendPage, SeoOptions } from "../shared";

export const logger = new Logger("vuepress-plugin-seo2");

export interface LocaleConfig {
  localePath: string;
  lang: string;
}

export const getLocales = (
  lang: string,
  locales: SiteLocaleConfig
): LocaleConfig[] =>
  Object.entries(locales)
    .map(([localePath, value]) => ({ localePath, lang: value.lang }))
    .filter(
      (item): item is LocaleConfig =>
        typeof item.lang === "string" && item.lang !== lang
    );

export const getCover = (
  { frontmatter }: ExtendPage,
  { hostname }: SeoOptions,
  { options: { base } }: App
): string | null => {
  const { banner, cover } = frontmatter;

  if (banner) {
    if (isAbsoluteUrl(banner)) return resolveUrl(hostname, base, banner);

    if (isUrl(banner)) return banner;
  }

  if (cover) {
    if (isAbsoluteUrl(cover)) return resolveUrl(hostname, base, cover);

    if (isUrl(cover)) return cover;
  }

  return null;
};

export const getImages = (
  { content }: ExtendPage,
  { hostname }: SeoOptions,
  { options: { base } }: App
): string[] => {
  const result = /!\[.*?\]\((.*?)\)/giu.exec(content);

  if (result) {
    return result
      .map(([, link]) => {
        if (isAbsoluteUrl(link)) return resolveUrl(hostname, base, link);

        if (isUrl(link)) return link;

        return null;
      })
      .filter((item): item is string => item !== null);
  }

  return [];
};

export const resolveUrl = (
  hostname: string,
  base: string,
  url: string
): string =>
  `${
    isLinkHttp(hostname)
      ? removeEndingSlash(hostname)
      : `https://${removeEndingSlash(hostname)}`
  }${base}${removeLeadingSlash(url)}`;

export const stripTags = (content = ""): string =>
  content
    // remove html tags
    .replace(/<\/?.+?\/?>/g, "");

export const md2text = (content?: string): string =>
  content
    ? stripTags(matter(content).content)
        // remove img
        .replace(/!\[(.*?)\]\(.*?\)/gm, "")
        // remove code blocks
        .replace(/```([\s\S]*?)```/g, "")
        // remove custom container end
        .replace(/^\s*:::\s*$/gm, "")
        // remove custom container start
        .replace(/^\s*:::\s*(.+?)(?:\s+(.*))?$/gm, "$2")
        // remove heading1
        .replace(/^# (.*)$/gm, "$1")
        // convert other headings to text
        .replace(/^#{1,6} (.*)$/gm, "$1")
        // convert unordered lists to text with comma
        .replace(/^\s*[-*+] (.*)$/gm, "$1; ")
        // convert blockquotes with quotes
        .replace(/^\s*>+(.*)$/gm, '"$1"')
        // convert links to text
        .replace(/(^|[^\\])\[(.*?)\]\(.*?\)/gm, "$1$2")
        // convert inline code
        .replace(/`{1,2}([^`])(.*?)`{1,2}/g, "$1$2")
        // just remove delete lines
        .replace(/~~(.*?)~~/g, "")
        // remove bold or italic
        .replace(/(^|[^\\])([*|_]{1,2})(.*?)([^\\])\2/gm, "$1$3$4")
        // remove html tags
        .replace(/<\/?.+?\/?>/g, "")
        // trim lines
        .split("\n")
        .map((line) => line.trim())
        .join("\n")
        // convert link breaks into spaces
        .replace(/(?:\r?\n)+/g, " ")
        // convert 2 or more spaces into 1
        .replace(/ +/g, " ")
        // trim
        .trim()
    : "";
