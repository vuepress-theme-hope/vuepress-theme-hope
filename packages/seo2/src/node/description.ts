import { getPageText, stripTags } from "vuepress-shared/node";

import { type ExtendPage } from "./typings/index.js";

export const generateDescription = (
  page: ExtendPage,
  autoDescription = true
): void => {
  // generate description
  if (!page.frontmatter.description && autoDescription) {
    if (page.data.excerpt) {
      page.frontmatter.description = stripTags(page.data.excerpt)
        // convert link breaks into spaces
        .replace(/(?:\r?\n)+/g, " ")
        // convert 2 or more spaces into 1
        .replace(/ +/g, " ")
        .trim();
    } else {
      const pageText = getPageText(page);

      page.frontmatter.description =
        pageText.length > 180 ? `${pageText.slice(0, 177)}...` : pageText;
    }

    page.data.autoDesc = true;
  }
};
