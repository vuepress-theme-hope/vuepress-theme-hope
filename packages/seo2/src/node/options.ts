import { type App, type HeadConfig, type Page } from "@vuepress/core";
import { type Author } from "vuepress-shared/node";

import {
  type ArticleSchema,
  type BlogPostingSchema,
  type ExtendPage,
  type SeoContent,
  type WebPageSchema,
} from "./typings/index.js";

export interface SeoOptions {
  /**
   * Deploy hostname
   *
   * 部署域名
   */
  hostname: string;

  /**
   * Default author
   *
   * 默认作者
   */
  author?: Author;

  /**
   * Content restrictions
   *
   * The age rating of the content, the format is `[int]+`, such as `"13+"`
   *
   * 内容分级情况
   *
   * 内容的年龄分级，格式为`[int]+`，如`"13+"`
   */
  restrictions?: string;

  /**
   * Whether generate description automatically
   *
   * 是否自动生成描述
   *
   * @default true
   */
  autoDescription?: boolean;

  /**
   * Fallback Image link when no image are found
   *
   * @description should be full or absolute links, probably your site favicon
   *
   * 当找不到图片时的回退图片链接
   *
   * @description 应为完整或绝对链接，你可以设置为站点图标
   */
  fallBackImage?: string;

  /**
   * Twitter username
   *
   * Twitter 用户名
   */
  twitterID?: string;

  /**
   * Whether the page is an article
   *
   * 页面是否是文章
   */
  isArticle?: <
    ExtraPageData extends Record<string | number | symbol, unknown> = Record<
      never,
      never
    >,
    ExtraPageFrontmatter extends Record<
      string | number | symbol,
      unknown
    > = Record<string, unknown>,
    ExtraPageFields extends Record<string | number | symbol, unknown> = Record<
      never,
      never
    >
  >(
    page: ExtendPage<ExtraPageData, ExtraPageFrontmatter, ExtraPageFields>
  ) => boolean;

  /**
   * Custom OGP Generator
   *
   * 自定义 OGP 生成器
   */
  ogp?: <
    ExtraPageData extends Record<string | number | symbol, unknown> = Record<
      never,
      never
    >,
    ExtraPageFrontmatter extends Record<
      string | number | symbol,
      unknown
    > = Record<string, unknown>,
    ExtraPageFields extends Record<string | number | symbol, unknown> = Record<
      never,
      never
    >
  >(
    /**
     * OGP Object inferred by plugin
     *
     * 由插件推断出的 OGP 对象
     */
    ogp: SeoContent,
    /**
     * Page Object
     *
     * 页面对象
     */
    page: ExtendPage<ExtraPageData, ExtraPageFrontmatter, ExtraPageFields>,
    /** VuePress App */
    app: App
  ) => SeoContent;

  /**
   * Custom JSON-LD Generator
   *
   * 自定义 JSON-LD 生成器
   */
  jsonLd?: <
    ExtraPageData extends Record<string | number | symbol, unknown> = Record<
      never,
      never
    >,
    ExtraPageFrontmatter extends Record<
      string | number | symbol,
      unknown
    > = Record<string, unknown>,
    ExtraPageFields extends Record<string | number | symbol, unknown> = Record<
      never,
      never
    >
  >(
    /**
     * JSON-LD Object inferred by plugin
     *
     * 由插件推断出的 JSON-LD 对象
     */
    jsonLD: ArticleSchema | BlogPostingSchema | WebPageSchema,
    /**
     * Page Object
     *
     * 页面对象
     */
    page: ExtendPage<ExtraPageData, ExtraPageFrontmatter, ExtraPageFields>,
    /** VuePress App */
    app: App
  ) => ArticleSchema | BlogPostingSchema | WebPageSchema;

  /**
   * Custom head tags
   *
   * 自定义 Head 标签
   */
  customHead?: <
    ExtraPageData extends Record<string | number | symbol, unknown> = Record<
      never,
      never
    >,
    ExtraPageFrontmatter extends Record<
      string | number | symbol,
      unknown
    > = Record<string, unknown>,
    ExtraPageFields extends Record<string | number | symbol, unknown> = Record<
      never,
      never
    >
  >(
    /**
     * Head tag config
     *
     * head 标签配置
     */
    head: HeadConfig[],
    /**
     * Page Object
     *
     * 页面对象
     */
    page: Page<ExtraPageData, ExtraPageFrontmatter, ExtraPageFields>,
    /** VuePress App */
    app: App
  ) => void;

  /**
   * Add canonical URL
   *
   * 添加首选地址
   */
  canonical?:
    | string
    | (<
        ExtraPageData extends Record<
          string | number | symbol,
          unknown
        > = Record<never, never>,
        ExtraPageFrontmatter extends Record<
          string | number | symbol,
          unknown
        > = Record<string, unknown>,
        ExtraPageFields extends Record<
          string | number | symbol,
          unknown
        > = Record<never, never>
      >(
        page: ExtendPage<ExtraPageData, ExtraPageFrontmatter, ExtraPageFields>
      ) => string | null);
}
