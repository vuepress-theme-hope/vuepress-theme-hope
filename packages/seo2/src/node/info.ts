/* eslint-disable @typescript-eslint/naming-convention */
import { type App } from "@vuepress/core";
import {
  isArray,
  isFunction,
  isString,
  removeEndingSlash,
} from "@vuepress/shared";
import { getAuthor, getDate } from "vuepress-shared/node";

import { type SeoOptions } from "./options.js";
import {
  type ArticleSchema,
  type BlogPostingSchema,
  type ExtendPage,
  type SeoContent,
  type WebPageSchema,
} from "./typings/index.js";
import { getAlternateInfo, getCover, getImages, resolveUrl } from "./utils.js";

export const getOGP = (
  page: ExtendPage,
  options: SeoOptions,
  app: App
): SeoContent => {
  const {
    isArticle = (page): boolean =>
      Boolean(page.filePathRelative && !page.frontmatter["home"]),
    author: globalAuthor,
  } = options;
  const {
    options: { base },
    siteData,
  } = app;
  const {
    frontmatter: {
      author: pageAuthor,
      time,
      date = time,
      tag,
      tags = <string[]>tag,
    },
    data: { git = {} },
  } = page;

  const title =
    siteData.locales[page.pathLocale]?.title ||
    siteData.title ||
    siteData.locales["/"]?.title ||
    "";
  const author =
    pageAuthor === false ? [] : getAuthor(pageAuthor || globalAuthor);
  const modifiedTime = git.updatedTime
    ? new Date(git.updatedTime).toISOString()
    : null;
  const articleTags = isArray(tags) ? tags : isString(tag) ? [tag] : [];
  const articleTitle = page.title;
  const cover = getCover(page, app, options);
  const images = getImages(page, app, options);
  const locales = getAlternateInfo(page, app);
  const publishedTime = getDate(date)?.value?.toISOString();

  const ogImage = cover || images[0] || options.fallBackImage || "";

  const defaultOGP: SeoContent = {
    "og:url": resolveUrl(options.hostname, base, page.path),
    "og:site_name": title,
    "og:title": articleTitle,
    "og:description": page.frontmatter.description || "",
    "og:type": isArticle(page) ? "article" : "website",
    "og:image": ogImage,
    "og:locale": page.lang,
    "og:locale:alternate": locales.map(({ lang }) => lang),
    ...(modifiedTime ? { "og:updated_time": modifiedTime } : {}),
    ...(options.restrictions
      ? { "og:restrictions:age": options.restrictions }
      : {}),

    ...(options.twitterID ? { "twitter:creator": options.twitterID } : {}),
    ...(ogImage
      ? {
          "twitter:card": "summary_large_image",
          "twitter:image:alt": articleTitle,
        }
      : {}),

    "article:author": author[0]?.name,
    "article:tag": articleTags,
    ...(publishedTime ? { "article:published_time": publishedTime } : {}),
    ...(modifiedTime ? { "article:modified_time": modifiedTime } : {}),
  };

  return defaultOGP;
};

export const getJSONLD = (
  page: ExtendPage,
  options: SeoOptions,
  app: App
): ArticleSchema | BlogPostingSchema | WebPageSchema => {
  const {
    isArticle = (page): boolean =>
      Boolean(page.filePathRelative && !page.frontmatter["home"]),
    author: globalAuthor,
  } = options;

  const {
    title,
    frontmatter: { author: pageAuthor, description, time, date = time },
    data: { git = {} },
  } = page;

  const author =
    pageAuthor === false ? [] : getAuthor(pageAuthor || globalAuthor);
  const datePublished = getDate(date)?.value?.toISOString();
  const dateModified = git.updatedTime
    ? new Date(git.updatedTime).toISOString()
    : null;
  const cover = getCover(page, app, options);
  const images = getImages(page, app, options);

  return isArticle(page)
    ? {
        "@context": "https://schema.org",
        "@type": "Article",
        headline: title,
        image: images.length ? images : [cover || options.fallBackImage || ""],
        datePublished,
        dateModified,
        author: author.map((item) => ({ "@type": "Person", ...item })),
      }
    : {
        "@context": "https://schema.org",
        "@type": "WebPage",
        name: title,
        ...(description ? { description } : {}),
      };
};

export const getCanonicalLink = (
  page: ExtendPage,
  options: SeoOptions
): string | null =>
  isFunction(options.canonical)
    ? options.canonical(page)
    : isString(options.canonical)
    ? `${removeEndingSlash(options.canonical)}${page.path}`
    : null;

export const getAlternateLinks = (
  page: ExtendPage,
  { hostname }: SeoOptions,
  app: App
): { lang: string; path: string }[] =>
  getAlternateInfo(page, app).map(({ lang, path }) => ({
    lang,
    path: resolveUrl(hostname, app.options.base, path),
  }));
