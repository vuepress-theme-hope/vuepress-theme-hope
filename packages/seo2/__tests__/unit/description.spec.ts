/* eslint-disable @typescript-eslint/naming-convention */
import { createBaseApp } from "@vuepress/core";
import { path } from "@vuepress/utils";
import { describe, expect, it } from "vitest";

import { emptyTheme } from "./__fixtures__/theme/empty.js";
import { seoPlugin } from "../../src/node";

const app = createBaseApp({
  bundler: {} as any,
  source: path.resolve(__dirname, "./__fixtures__/src"),
  theme: emptyTheme,
  locales: {
    "/": {
      lang: "en-US",
    },
    "/zh/": {
      lang: "zh-CN",
    },
  },
  plugins: [
    seoPlugin({
      hostname: "https://exmaple.com",
      canonical: "https://exmaple.com",
    }),
  ],
});

await app.init();

describe("Should generate seo information", () => {
  it("Should contain basic properties", () => {
    app.pages.forEach(({ frontmatter }) => {
      expect(frontmatter.head).toMatchSnapshot();
    });
  });
});
