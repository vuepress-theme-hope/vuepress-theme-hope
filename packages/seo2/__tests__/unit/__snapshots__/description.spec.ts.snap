// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should generate seo information > Should contain basic properties 1`] = `
[
  [
    "link",
    {
      "href": "https://exmaple.com/",
      "rel": "canonical",
    },
  ],
  [
    "link",
    {
      "href": "https://exmaple.com/zh/",
      "hreflang": "zh-cn",
      "rel": "alternate",
    },
  ],
  [
    "meta",
    {
      "content": "https://exmaple.com/",
      "property": "og:url",
    },
  ],
  [
    "meta",
    {
      "content": "This is a home page.",
      "property": "og:description",
    },
  ],
  [
    "meta",
    {
      "content": "website",
      "property": "og:type",
    },
  ],
  [
    "meta",
    {
      "content": "en-US",
      "property": "og:locale",
    },
  ],
  [
    "meta",
    {
      "content": "zh-CN",
      "property": "og:locale:alternate",
    },
  ],
  [
    "script",
    {
      "type": "application/ld+json",
    },
    "{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"WebPage\\",\\"name\\":\\"\\",\\"description\\":\\"This is a home page.\\"}",
  ],
]
`;

exports[`Should generate seo information > Should contain basic properties 2`] = `
[
  [
    "link",
    {
      "href": "https://exmaple.com/description.html",
      "rel": "canonical",
    },
  ],
  [
    "link",
    {
      "href": "https://exmaple.com/zh/description.html",
      "hreflang": "zh-cn",
      "rel": "alternate",
    },
  ],
  [
    "meta",
    {
      "content": "https://exmaple.com/description.html",
      "property": "og:url",
    },
  ],
  [
    "meta",
    {
      "content": "Description Test",
      "property": "og:title",
    },
  ],
  [
    "meta",
    {
      "content": "This page is used to test the description of the page.",
      "property": "og:description",
    },
  ],
  [
    "meta",
    {
      "content": "article",
      "property": "og:type",
    },
  ],
  [
    "meta",
    {
      "content": "en-US",
      "property": "og:locale",
    },
  ],
  [
    "meta",
    {
      "content": "zh-CN",
      "property": "og:locale:alternate",
    },
  ],
  [
    "script",
    {
      "type": "application/ld+json",
    },
    "{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"Article\\",\\"headline\\":\\"Description Test\\",\\"image\\":[\\"\\"],\\"dateModified\\":null,\\"author\\":[]}",
  ],
]
`;

exports[`Should generate seo information > Should contain basic properties 3`] = `
[
  [
    "link",
    {
      "href": "https://exmaple.com/example.html",
      "rel": "canonical",
    },
  ],
  [
    "link",
    {
      "href": "https://exmaple.com/zh/example.html",
      "hreflang": "zh-cn",
      "rel": "alternate",
    },
  ],
  [
    "meta",
    {
      "content": "https://exmaple.com/example.html",
      "property": "og:url",
    },
  ],
  [
    "meta",
    {
      "content": "Heading 1",
      "property": "og:title",
    },
  ],
  [
    "meta",
    {
      "content": "Heading 2 Heading 3 Heading 4 Heading 5 Heading 6 Text This sentence has bold、italic and style text. Paragraph This is a paragraph. This is another paragraph. Line Break I would...",
      "property": "og:description",
    },
  ],
  [
    "meta",
    {
      "content": "article",
      "property": "og:type",
    },
  ],
  [
    "meta",
    {
      "content": "en-US",
      "property": "og:locale",
    },
  ],
  [
    "meta",
    {
      "content": "zh-CN",
      "property": "og:locale:alternate",
    },
  ],
  [
    "script",
    {
      "type": "application/ld+json",
    },
    "{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"Article\\",\\"headline\\":\\"Heading 1\\",\\"image\\":[\\"\\"],\\"dateModified\\":null,\\"author\\":[]}",
  ],
]
`;

exports[`Should generate seo information > Should contain basic properties 4`] = `
[
  [
    "link",
    {
      "href": "https://exmaple.com/zh/",
      "rel": "canonical",
    },
  ],
  [
    "link",
    {
      "href": "https://exmaple.com/",
      "hreflang": "en-us",
      "rel": "alternate",
    },
  ],
  [
    "meta",
    {
      "content": "https://exmaple.com/zh/",
      "property": "og:url",
    },
  ],
  [
    "meta",
    {
      "content": "这是一个主页。",
      "property": "og:description",
    },
  ],
  [
    "meta",
    {
      "content": "website",
      "property": "og:type",
    },
  ],
  [
    "meta",
    {
      "content": "zh-CN",
      "property": "og:locale",
    },
  ],
  [
    "meta",
    {
      "content": "en-US",
      "property": "og:locale:alternate",
    },
  ],
  [
    "script",
    {
      "type": "application/ld+json",
    },
    "{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"WebPage\\",\\"name\\":\\"\\",\\"description\\":\\"这是一个主页。\\"}",
  ],
]
`;

exports[`Should generate seo information > Should contain basic properties 5`] = `
[
  [
    "link",
    {
      "href": "https://exmaple.com/zh/description.html",
      "rel": "canonical",
    },
  ],
  [
    "link",
    {
      "href": "https://exmaple.com/description.html",
      "hreflang": "en-us",
      "rel": "alternate",
    },
  ],
  [
    "meta",
    {
      "content": "https://exmaple.com/zh/description.html",
      "property": "og:url",
    },
  ],
  [
    "meta",
    {
      "content": "描述测试",
      "property": "og:title",
    },
  ],
  [
    "meta",
    {
      "content": "这个页面用于测试页面描述",
      "property": "og:description",
    },
  ],
  [
    "meta",
    {
      "content": "article",
      "property": "og:type",
    },
  ],
  [
    "meta",
    {
      "content": "zh-CN",
      "property": "og:locale",
    },
  ],
  [
    "meta",
    {
      "content": "en-US",
      "property": "og:locale:alternate",
    },
  ],
  [
    "script",
    {
      "type": "application/ld+json",
    },
    "{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"Article\\",\\"headline\\":\\"描述测试\\",\\"image\\":[\\"\\"],\\"dateModified\\":null,\\"author\\":[]}",
  ],
]
`;

exports[`Should generate seo information > Should contain basic properties 6`] = `
[
  [
    "link",
    {
      "href": "https://exmaple.com/zh/example.html",
      "rel": "canonical",
    },
  ],
  [
    "link",
    {
      "href": "https://exmaple.com/example.html",
      "hreflang": "en-us",
      "rel": "alternate",
    },
  ],
  [
    "meta",
    {
      "content": "https://exmaple.com/zh/example.html",
      "property": "og:url",
    },
  ],
  [
    "meta",
    {
      "content": "一级标题",
      "property": "og:title",
    },
  ],
  [
    "meta",
    {
      "content": "二级标题 三级标题 四级标题 五级标题 六级标题 Text 这句话里拥有加粗、倾斜和 段落 这是一个段落。 这是另一个段落。 换行 这是一句话不过我要在这里 换行 上方的代码中 这里 后面有两个空格 引用 引用也可以连用 可以添加额外的大于号制造更深的引用 列表 无序列表 无序列表项; 无序列表项; 列表中的列表项; 更多的列表项; 更多的列表项; 更...",
      "property": "og:description",
    },
  ],
  [
    "meta",
    {
      "content": "article",
      "property": "og:type",
    },
  ],
  [
    "meta",
    {
      "content": "zh-CN",
      "property": "og:locale",
    },
  ],
  [
    "meta",
    {
      "content": "en-US",
      "property": "og:locale:alternate",
    },
  ],
  [
    "script",
    {
      "type": "application/ld+json",
    },
    "{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"Article\\",\\"headline\\":\\"一级标题\\",\\"image\\":[\\"\\"],\\"dateModified\\":null,\\"author\\":[]}",
  ],
]
`;

exports[`Should generate seo information > Should contain basic properties 7`] = `
[
  [
    "link",
    {
      "href": "https://exmaple.com/404.html",
      "rel": "canonical",
    },
  ],
  [
    "meta",
    {
      "content": "https://exmaple.com/404.html",
      "property": "og:url",
    },
  ],
  [
    "meta",
    {
      "content": "website",
      "property": "og:type",
    },
  ],
  [
    "meta",
    {
      "content": "en-US",
      "property": "og:locale",
    },
  ],
  [
    "script",
    {
      "type": "application/ld+json",
    },
    "{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"WebPage\\",\\"name\\":\\"\\"}",
  ],
]
`;
